088-rebalance-progress-status.md

```diff
@@ -0,0 +1,374 @@
+
+Example of the Executor State JSON payload during an inter-broker balance:
+```json
+{
+  "abortingPartitions": 0,
+  "averageConcurrentPartitionMovementsPerBroker": 5,
+  "finishedDataMovement": 0,
+  "maximumConcurrentPartitionMovementsPerBroker": 5,
+  "minimumConcurrentPartitionMovementsPerBroker": 5,
+  "numFinishedPartitionMovements": 0,
+  "numInProgressPartitionMovements": 0,
+  "numPendingPartitionMovements": 20,
+  "numTotalPartitionMovements": 20,
+  "state": "INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
+  "totalDataToMove": 0,
+  "triggeredSelfHealingTaskId": "",
+  "triggeredTaskReason": "No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
+  "triggeredUserTaskId": "0230d401-6a36-430e-9858-fac8f2edde93"
+}
+```
+
+The formulas used to calculate field value per  `KafkaRebalance`  state:
+
+#### State: `Ready`
+
+The rebalance has not started yet so no data surrounding partition movement is available yet, so we set the field to an empty JSON object.
+
+#### State: `Rebalancing`
+
+The field will be assigned the "non-verbose" JSON payload of the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of the Cruise Control REST API.
+
+#### State: `Stopped`
+
+Once a rebalance has been stopped, it cannot be resumed.
+However, the `executorState` information from when before the rebalance was stopped may still be of value to users.
+Therefore, we reuse the same value of `executorState` from the previous update.
+
+#### State: `Ready`
+
+The rebalance is complete in this state so we hardcode the value to an empty JSON object.
+
+### Progress Update Cadence
+
+For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.
+
+In the event that Cruise Control runs into an error when rebalancing, the operator will transition the `KafkaRebalance` resource to the `NotReady` state, remove the `progress` section, and delete the progress `ConfigMap`.
+In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.
+
+When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  ...
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Warning
```

Comment:
Should we call out this property to highlight and distinguish between `NotReady` and (new?) `Warning`?

Reply:
The type of condition in the status could be any one of the `KafkaRebalance` states, it could also be `Warning`. Why would the `NotReady` type be tied/associated to the `Warning` type?

Reply:
@kyguy I think this is a good point from Paul that confused me as well -> you should have one of the regular states included in the example if they would accompany the warning.

Reply:
Sorry Paul, I misunderstood your suggestion, I'll add a `KafkaRebalance` state condition to the example to draw a distinguish between the `KafkaRebalance` state condition and the `Warning` condition.